### PR TITLE
[libc++] Remove workaround which allows setting _LIBCPP_OVERRIDABLE_FUNC_VIS externally

### PR DESCRIPTION
**Pull request by philnik777**
`-fvisibility-global-new-delete` has been added in Clang 18, so there is no more need to specify the visibility of new/delete via libc++-internal macros.

Some people used a custom new/delete, which requires them to have default visibility, but didn't want to leak any symbols otherwise. Since the new/delete visibility can now be controlled by a compiler flag, there is no reason to set it with out macro.

https://reviews.llvm.org/D128007 originally tried to remove the option to override the macro, but was partially reverted because people actually set this specific macro.

### CODE REVIEW

**Comment by philnik777** [comment]
CC @rnk

**Comment by ldionne** [comment]
CC @llvm/libcxx-vendors 

@philnik777 Can you please add a short explanation of the history of this flag? I think you removed it and that broke a use case, and now it can be removed. What broke when you originally removed it?

**Comment by philnik777** [comment]
> CC @llvm/libcxx-vendors
> 
> @philnik777 Can you please add a short explanation of the history of this flag? I think you removed it and that broke a use case, and now it can be removed. What broke when you originally removed it?

Some people used a custom new/delete, which requires them to have default visibility, but didn't want to leak any symbols otherwise. Since the new/delete visibility can now be controlled by a compiler flag, there is no reason to set it with out macro.

**Comment by rnk** [comment]
I think Louis is asking for more context links, which would be helpful for reviewers and future readers. My best efforts to find the history turned up https://reviews.llvm.org/D129768, but I couldn't find any more useful history on the Chromium side. It looks like Chromium no longer users this macro, but I can't figure out if it ever did or when that changed.

**Review by ldionne — APPROVED** [review]
LGTM, but please make sure to include the explanation for the history of this change in the PR description (i.e. the commit message once landed).

**Comment by nico** [comment]
Heads-up: This broke a bunch of tests for us (chromium) that we're still analyzing.

However, trying `-fvisibility-global-new-delete` on a small demo shows that using it basically doesn't work:

```
% cat test-attrib.cc
#include <new>

int main() {
  int* a = new int;
  return *a;
}

thakis@MacBook-Pro llvm-project % out/gn/bin/clang++ test-attrib.cc -isysroot $(xcrun -show-sdk-path)  -fuse-ld=lld --target=x86_64-unknown-linux --sysroot ~/src/chrome/src/build/linux/debian_bullseye_amd64-sysroot -fvisibility-global-new-delete=force-hidden -stdlib=libc++
In file included from test-attrib.cc:1:
/Users/thakis/src/llvm-project/out/gn/bin/../include/c++/v1/new:212:19: error: visibility does not match previous declaration
  212 | [[__nodiscard__]] _LIBCPP_OVERRIDABLE_FUNC_VIS void* operator new(std::size_t __sz) _THROW_BAD_ALLOC;
      |                   ^
/Users/thakis/src/llvm-project/out/gn/bin/../include/c++/v1/__config:411:42: note: expanded from macro '_LIBCPP_OVERRIDABLE_FUNC_VIS'
  411 | #    define _LIBCPP_OVERRIDABLE_FUNC_VIS _LIBCPP_VISIBILITY("default")
      |                                          ^
/Users/thakis/src/llvm-project/out/gn/bin/../include/c++/v1/__config:401:54: note: expanded from macro '_LIBCPP_VISIBILITY'
  401 | #      define _LIBCPP_VISIBILITY(vis) __attribute__((__visibility__(vis)))
      |                                                      ^
note: previous attribute is here
```

I.e. all it does is give you a "visibility doesn't match" error.

Am I holding it wrong? How is `-fvisibility-global-new-delete=force-hidden` to be used with this?

Also, libcxx/docs/DesignDocs/VisibilityMacros.rst still documents `_LIBCPP_OVERRIDABLE_FUNC_VIS`.

**Comment by nico** [comment]
(ps: Sorry it took us a while to notice; we were blocked on libc++ adding a dep on LLVM's libc for a while and had to sort that out first.)

**Comment by rnk** [comment]
I guess I misspoke, searching again for `_LIBCPP_OVERRIDABLE_FUNC_VIS` points to the relevant long comment:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/buildtools/third_party/libc++/BUILD.gn;l=213?q=_LIBCPP_OVERRIDABLE_FUNC_VIS&ss=chromium&start=41

So, no, Chromium never migrated to this new flag.

**Comment by atetubou** [comment]
From my investigation, I think `-fvisibility-global-new-delete` isn't applied to `operator new` with `std::nothrow_t` arg around
https://github.com/llvm/llvm-project/blob/927032807dfdca5d94eb0a8707d38b605e95e407/clang/lib/Sema/SemaExprCXX.cpp#L3116
While `_LIBCPP_OVERRIDABLE_FUNC_VIS` is used for `operator new` with `std::nothrow_t`.
e.g. https://github.com/llvm/llvm-project/blob/738a047ed6380efdfd4b64968881675347d7f915/libcxx/include/new#L216

**Comment by nico** [comment]
I filed https://github.com/llvm/llvm-project/issues/117571 asking for a revert of this PR.

**Pull request merged**
